### PR TITLE
Add StackExchange.Utils.Configuration

### PR DESCRIPTION
**Pull request by deanward81**
## Overview
This PR adds a helper library that performs substitution and prefixing for `IConfiguration`-based configuration sources. It allows a value in the configuration tree to reference other values within the configuration system using a placeholder syntax `${key}` or `${nested:key}`. In addition, prefixing allows a "namespace" prefix to be applied to subset of configuration, making it possible to segment configuration into logical areas.

This is particularly useful for storing secrets in a different, secure location but in a way that it makes it easy to compose configuration like connection strings without dealing with it inside the application. E.g. consider the following JSON:

**appsettings.json**
```json
{
    "ConnectionStrings": {
        "Database": "Server=srv01;Database=db01;User ID=${secrets:Database:UserId};Password=${secrets:Database:Password}"
    }
}
```

**secrets.json**
```json
{
    "Database": {
        "UserId": "User123",
        "Password": "Password123!"
    }
}
```

This instructs the substitution provider to lookup the keys `secrets:Database:UserId` and `secrets:Database:Password` from the configuration system and replaces them in the value returned for `ConnectionStrings:Database`.

To support this a `ConfigurationBuilder` is configured as follows:

```c#
var configuration = new ConfigurationBuilder()
    .WithPrefix(
        "secrets",
        // everything in this configuration builder will be prefixed with "secrets:"
        c => c.AddJsonFile("secrets.json")
    )
    .WithSubstitution(
        // values in this configuration builder will have substitutions
        // replaced prior to being returned to the caller
        c => c.AddJsonFile("appsettings.json")
    )
    .Build();
```

Here we're loading a JSON file called `secrets.json` (it could equally be any source supported by the `IConfiguration` system) and prefixing it with `secrets:`. Then, we load `appsettings.json` with support for substitutions. If a caller asks the `IConfiguration` that is produced for a connection string:

```c#
var connectionString = configuration.GetConnectionString("Database");
```

That value will be returned as `Server=srv01;Database=db01;User ID=User123;Password=Password123!`.

## Other Changes

- Build scripts updated to build all projects
- Builds scripts updated to support Linux/MacOS
- `.gitignore` tweaks for Rider
- Move to `LangVersion = latest`
- `version.json` in each packages directories
- Use `Microsoft.NETFramework.ReferenceAssemblies` for net462 builds on Linux

### CODE REVIEW

File: build.ps1 [hunk 1]
```diff
@@ -10,9 +10,10 @@ Write-Host "  CreatePackages: $CreatePackages"
 Write-Host "  RunTests: $RunTests"
 Write-Host "  dotnet --version:" (dotnet --version)
 
-$packageOutputFolder = "$PSScriptRoot\.nupkgs"
+$packageOutputFolder = Join-Path $PSScriptRoot ".nupkgs"
 $projectsToBuild =
-    'StackExchange.Utils.Http'
+    'StackExchange.Utils.Http',
+    'StackExchange.Utils.Configuration'
```
Comment (NickCraver) [inline build.ps1]: We should probably use the Build.csproj version of this now - see https://github.com/StackExchange/StackExchange.Redis/blob/main/build.ps1 - means we no longer list the projects at all!

**Comment by NickCraver** [comment]
Looking through now!

File: src/StackExchange.Utils.Configuration/CompositeConfigurationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            if (_initialLoad)
+            {
+                _initialLoad = false;
+                return;
+            }
+            ConfigurationRoot.Reload();
+        }
+
+        public void Set(string key, string value) => throw new NotSupportedException();
```
Comment (NickCraver) [inline src/StackExchange.Utils.Configuration/CompositeConfigurationProvider.cs]: Intended to be unsupported going forward?

File: src/StackExchange.Utils.Configuration/ConfigurationBuilderExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        /// <summary>
+        /// Wraps a child <see cref="IConfigurationBuilder"/> with support for substituting values
+        /// of the form '${secrets:Database:Password}' with configuration from elsewhere in the
+        /// configuration system. The key in braces is looked and used to rewrite the value returned
+        /// from configuration. 
+        /// </summary>
+        public static IConfigurationBuilder WithSubstitution(this IConfigurationBuilder builder, Action<IConfigurationBuilder> action)
+        {
+            var childBuilder = new ConfigurationBuilder();
+            action(childBuilder);
```
Comment (NickCraver) [inline src/StackExchange.Utils.Configuration/ConfigurationBuilderExtensions.cs:23]: `?? throw` null check on `action` would be good

File: src/StackExchange.Utils.Configuration/ConfigurationBuilderExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+        /// <summary>
+        /// Wraps a child <see cref="IConfigurationBuilder"/> with support for a prefix on all keys
+        /// contained within it. This can be used to effectively namespace a set of configuration values
+        /// and is useful when used in combination with <see cref="WithSubstitution"/>.
+        /// </summary>
+        public static IConfigurationBuilder WithPrefix(this IConfigurationBuilder builder, string prefix, Action<IConfigurationBuilder> action)
+        {
+            var childBuilder = new ConfigurationBuilder();
+            action(childBuilder);
```
Comment (NickCraver) [inline src/StackExchange.Utils.Configuration/ConfigurationBuilderExtensions.cs:39]: Same as above

File: src/StackExchange.Utils.Configuration/SubstitutionHelper.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        private class CycleDetector : ICycleDetector
+        {
+            private readonly HashSet<string> _values = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            public void Add(string key) => _values.Add(key);
+
+            public void Remove(string key) => _values.Remove(key);
+
+            public bool Contains(string key) => _values.Contains(key);
+        }
```
Comment (NickCraver) [inline src/StackExchange.Utils.Configuration/SubstitutionHelper.cs:147]: This works, but FWIW could simplify a bit with a `Stack<string>` in the `AsyncLocal` as well I think.

File: tests/StackExchange.Utils.Tests/StackExchange.Utils.Tests.csproj [hunk 6]
```diff
@@ -1,16 +1,12 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (NickCraver) [inline tests/StackExchange.Utils.Tests/StackExchange.Utils.Tests.csproj:3]: Need to remove the line below if we're simplifying to 1 TFM here.

File: tests/StackExchange.Utils.Tests/StackExchange.Utils.Tests.csproj [hunk 7]
```diff
@@ -1,16 +1,12 @@
...
     <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
   </PropertyGroup>
   <ItemGroup>
-    <ProjectReference Include="../../src/StackExchange.Utils.Http/StackExchange.Utils.Http.csproj" />
-  </ItemGroup>
-
-  <PropertyGroup Condition="'$(TargetFramework)'=='netcoreapp3.1'">
-    <DefineConstants>$(DefineConstants);KESTREL</DefineConstants>
```
Comment (NickCraver) [inline tests/StackExchange.Utils.Tests/StackExchange.Utils.Tests.csproj:10]: If removing this (:+1:), need to remove the #if KESTREL from tests we wrap.

**Review by NickCraver — CHANGES_REQUESTED** [review]
Found a few bits - flagged inline!

File: src/StackExchange.Utils.Configuration/SubstitutionHelper.cs (hunk 5 above)
Comment (deanward81) [reply src/StackExchange.Utils.Configuration/SubstitutionHelper.cs:147]: Might be missing something here, but I'm not sure a `Stack<string>` works here? We're trying to detect cycles so if:

`keyA = ${keyB}`
`keyB = ${keyC}`
`keyC = ${keyA}`

We'd need to go enumerate all values in the stack? Only reason this is abstracted out to an interface is that we can't really do cycle detection for the simple value (not provider-based) replacement case so there's a no-op implementation

File: src/StackExchange.Utils.Configuration/CompositeConfigurationProvider.cs (hunk 2 above)
Comment (deanward81) [reply src/StackExchange.Utils.Configuration/CompositeConfigurationProvider.cs]: Good question! Not sure, mulling it over :)

Comment (deanward81) [reply src/StackExchange.Utils.Configuration/CompositeConfigurationProvider.cs]: Turns out it's nice and simple to implement, just pushed it

**Comment by NickCraver** [comment]
Pulled for a run locally, HTTP/2 tests are broken...not sure of current goal state here. I see the certificate was removed in the repo, but it's needed still - what am I missing here?
```
StackExchange.Utils.Tests.Http2Tests.UsesVersion.UsesVersion(protocols: Http1, tls: False, allowUnencryptedHttp2: False, specified: "1.1", expectedVersion: "1.1", expectedResponse: "HTTP/1.1", failure: False)

System.AggregateException: One or more errors occurred. (One or more errors occurred. (Could not find file '/Users/ncraver/Library/Application Support...

System.AggregateException
One or more errors occurred. (One or more errors occurred. (Could not find file '/Users/ncraver/Library/Application Support/JetBrains/Toolbox/apps/Rider/ch-0/202.6397.19/Rider EAP.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/certificate.pfx'.)) (The following constructor parameters did not have matching fixture data: Http2Server server)
  Exception doesn't have a stacktrace

System.AggregateException
One or more errors occurred. (Could not find file '/Users/ncraver/Library/Application Support/JetBrains/Toolbox/apps/Rider/ch-0/202.6397.19/Rider EAP.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/certificate.pfx'.)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at StackExchange.Utils.Tests.Http2Tests.Http2Server..ctor() in /Users/ncraver/git/StackExchange/StackExchange.Utils/tests/StackExchange.Utils.Tests/Http2Tests.cs:line 189

System.IO.FileNotFoundException
Could not find file '/Users/ncraver/Library/Application Support/JetBrains/Toolbox/apps/Rider/ch-0/202.6397.19/Rider EAP.app/Contents/lib/ReSharperHost/TestRunner/netcoreapp3.0/certificate.pfx'.
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode)
   at System.IO.FileStream.OpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize)
   at System.IO.File.ReadAllBytes(String path)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(String fileName, String password, X509KeyStorageFlags keyStorageFlags)
   at Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps(ListenOptions listenOptions, String fileName, String password)
   at StackExchange.Utils.Tests.Http2Tests.Http2Server.<>c.<.ctor>b__4_5(ListenOptions listenOptions) in /Users/ncraver/git/StackExchange/StackExchange.Utils/tests/StackExchange.Utils.Tests/Http2Tests.cs:line 163
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.ListenLocalhost(Int32 port, Action`1 configure)
   at StackExchange.Utils.Tests.Http2Tests.Http2Server.<.ctor>b__4_0(KestrelServerOptions options) in /Users/ncraver/git/StackExchange/StackExchange.Utils/tests/StackExchange.Utils.Tests/Http2Tests.cs:line 160
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.CreateServiceContext(IOptions`1 options, ILoggerFactory loggerFactory)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer..ctor(IOptions`1 options, ITransportFactory transportFactory, ILoggerFactory loggerFactory)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.EnsureServer()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
   at Microsoft.AspNetCore.Hosting.Internal.WebHost.StartAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(IWebHost host, CancellationToken token, String shutdownMessage)
   at Microsoft.AspNetCore.Hosting.WebHostExtensions.RunAsync(IWebHost host, CancellationToken token)

Xunit.Sdk.TestClassException
The following constructor parameters did not have matching fixture data: Http2Server server
  Exception doesn't have a stacktrace
```
Screen Shot 2020-07-26 at 4 42 11 PM

**Comment by deanward81** [comment]
@NickCraver smells like a Rider EAP bug - tests run on my machine running Rider's current release and using `dotnet test`...

**Comment by deanward81** [comment]
Certificate is still there (modified to add a password - MacOS doesn't like unencrypted PKCS12 files), something jacked up with your local working tree?

**Comment by NickCraver** [comment]
hmm indeed `dotnet test` is fine:
```
> dotnet test Build.csproj

Test run for /Users/ncraver/git/StackExchange/StackExchange.Utils/tests/StackExchange.Utils.Tests/bin/netcoreapp3.1/StackExchange.Utils.Tests.dll(.NETCoreApp,Version=v3.1)
Microsoft (R) Test Execution Command Line Tool Version 16.6.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
Hosting environment: Production
Content root path: /Users/ncraver/git/StackExchange/StackExchange.Utils/tests/StackExchange.Utils.Tests/bin/netcoreapp3.1/
Now listening on: http://localhost:10123
Now listening on: http://localhost:10124
Now listening on: http://localhost:10125
Now listening on: https://localhost:10126
Now listening on: https://localhost:10127
Now listening on: https://localhost:10128
Application started. Press Ctrl+C to shut down.
[xUnit.net 00:00:01.45]     StackExchange.Utils.Tests.Http2Tests.UsesVersion(protocols: Http2, tls: True, allowUnencryptedHttp2: False, specified: "2.0", expectedVersion: "2.0", expectedResponse: "HTTP/2", failure: False) [SKIP]
  ! StackExchange.Utils.Tests.Http2Tests.UsesVersion(protocols: Http2, tls: True, allowUnencryptedHttp2: False, specified: "2.0", expectedVersion: "2.0", expectedResponse: "HTTP/2", failure: False) [1ms]
[xUnit.net 00:00:01.79]     StackExchange.Utils.Tests.Http2Tests.UsesVersion(protocols: Http1AndHttp2, tls: True, allowUnencryptedHttp2: False, specified: "2.0", expectedVersion: "2.0", expectedResponse: "HTTP/2", failure: False) [SKIP]
  ! StackExchange.Utils.Tests.Http2Tests.UsesVersion(protocols: Http1AndHttp2, tls: True, allowUnencryptedHttp2: False, specified: "2.0", expectedVersion: "2.0", expectedResponse: "HTTP/2", failure: False) [1ms]
Application is shutting down...

Test Run Successful.
Total tests: 73
     Passed: 71
    Skipped: 2
 Total time: 7.2361 Seconds
```
...will file an issue with the Rider folks!

**Review by NickCraver — APPROVED** [review]
Looking good!

**Pull request merged**